src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -968,12 +990,22 @@ private void WriteHeaders(HttpRequestMessage request)
 
         private async ValueTask<Http2Stream> SendHeadersAsync(HttpRequestMessage request, CancellationToken cancellationToken, bool mustFlush = false)
         {
-            // Ensure we don't exceed the max concurrent streams setting.
-            await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
-
-            // We serialize usage of the header encoder and the header buffer separately from the
-            // write lock
-            await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                // Ensure we don't exceed the max concurrent streams setting.
+                await _concurrentStreams.RequestCreditAsync(1, cancellationToken).ConfigureAwait(false);
+                // We serialize usage of the header encoder and the header buffer separately from the
+                // write lock
+                await _headerSerializationLock.WaitAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (ObjectDisposedException)
+            {
+                // We have race condition between receiving GOAWAY and processing requests.
+                // Throw a retryable request exception if this is not result of some other error.
+                // This will cause retry logic to kick in and perform another connection attempt.
+                // The user should never see this exception.  Same logic lives in AddStream.
+                throw _abortException ?? new HttpRequestException(null, null, allowRetry: true);
```

Comment:
It'd be better not to do `throw _abortException`; doing that will overwrite the stack trace and watson bucket information in the exception with the current details.  If you instead do `ExceptionDispatchInfo.Throw(_abortException)` then the previous details will be retained and the current location will be appended to the exception rather than overwriting its state.